### PR TITLE
Multi-Memories Lowering Pass

### PR DESCRIPTION
**Pull request by ashleynh**
Adds a multi-memories lowering pass that will create a single combined memory from the memories added to the module. This pass assumes that each memory is configured the same (type, shared).

This pass also:
- replaces existing memory.size instructions with a custom function that returns the size of each memory as if they existed independently
- replaces existing memory.grow instructions with a custom function, using global offsets to track the page size of each memory so data doesn't overlap in the singled combined memory
- adjusts the offsets of active data segments
- adjusts the offsets of Loads/Stores

### CODE REVIEW

File: test/lit/passes/multi-memory-lowering.wast [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+(module
+  (memory $memory1 1)
+  (memory $memory2 2)
+  (data (memory $memory1) (i32.const 0) "a")
+  (data (memory $memory2) (i32.const 1) "123")
+)
+
+;; CHECK: (module
+ ;; CHECK-NEXT:(type $none_=>_i32 (func (result i32)))
```
Comment (tlively) [inline test/lit/passes/multi-memory-lowering.wast:12]: Did update_lit_checks.py emit these unaligned checks? This seems like a bug in the script if so. It looks like CHECK prefix and the whitespace that is supposed to come after it have been swapped.

File: src/passes/MultiMemoryLowering.cpp [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        Index idx = iter->second - 1;
+        curr->ptr = builder.makeGlobalGet(parent.offsetGlobalNames[idx],
+                                          parent.pointerType);
+      }
+    };
+    Replacer(*this, *wasm).run(runner, wasm);
+  }
+
+  void addCombinedMemory() {
+    // We are assuming that each memory is configured the same as the first
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:144]: Perhaps we should assert on them all having the same index type?

File: src/passes/MultiMemoryLowering.cpp [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+  }
+
+  void addOffsetGlobals() {
+    auto addGlobal = [&](Name name, size_t offset) {
+      auto global = Builder::makeGlobal(
+        name,
+        pointerType,
+        Builder(*wasm).makeConst(pointerType == Type::i32
+                                   ? Literal(int32_t(offset))
+                                   : Literal(int64_t(offset))),
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:169]: This can use `Literal::makeFromInt64(offset, pointerType)`

File: src/passes/MultiMemoryLowering.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+      auto iter = memoryIdxMap.find(dataSegment->memory);
+      assert(iter != memoryIdxMap.end());
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (iter->second != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = offsetGlobalNames[iter->second - 1];
+        dataSegment->offset = builder.makeBinary(
+          pointerType == Type::i32 ? AddInt32 : AddInt64,
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:205]: This can use `Abstract::getBinary(pointerType, Abstract::Add)`

File: src/passes/MultiMemoryLowering.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  }
+
+  // This function replaces memory.size instructions with a function that can
+  // return the size of each memory as if each was discrete and separate.
+  std::unique_ptr<Function> memorySize(Index memIdx) {
+    Builder builder(*wasm);
+    Name functionName = Names::getValidFunctionName(*wasm, "multi_memory_size");
+    auto function = Builder::makeFunction(
+      functionName, Signature(Type::none, pointerType), {});
+    function->hasExplicitName = true;
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:302]: Why is this needed? I suspect it can be removed.

**Review by kripken — COMMENTED** [review]
Nice work! Some comments on the code (haven't gotten to the test yet).

File: src/passes/MultiMemoryLowering.cpp [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        addGlobal(name, offsetRunningTotal);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>());
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:195]:
```suggestion
      assert(dataSegment->offset->is<Const>() && "TODO: handle non-const segment offsets");
```

File: src/passes/MultiMemoryLowering.cpp [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+      adjustOffsetNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> adjustOffsetGlobals(Index memIdx) {
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:236]: How about calling this `memoryGrow`, like how you call the helper function for memory.size `memorySize`?

File: src/passes/MultiMemoryLowering.cpp [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    auto getPageDelta = [&]() { return builder.makeLocalGet(0, pointerType); };
+    Expression* functionBody;
+
+    Index returnLocal = Builder::addVar(function.get(), pointerType);
+    functionBody = builder.blockify(builder.makeLocalSet(
+      returnLocal, builder.makeCall(memorySizeNames[memIdx], {}, pointerType)));
+
+    // If we are not growing the last memory, then we need to copy data,
+    // shifting it over to accomodate the increase from page_delta
+    if (memIdx != offsetGlobalNames.size()) {
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:253]: Since this also depends on the detail that `offsetGlobalNames` contains N-1 memories, it would be good to make a helper function like `isLastMemory` for this.

File: src/passes/MultiMemoryLowering.cpp [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+    function->hasExplicitName = true;
+    function->setLocalName(0, "page_delta");
+    auto getPageDelta = [&]() { return builder.makeLocalGet(0, pointerType); };
+    Expression* functionBody;
+
+    Index returnLocal = Builder::addVar(function.get(), pointerType);
+    functionBody = builder.blockify(builder.makeLocalSet(
+      returnLocal, builder.makeCall(memorySizeNames[memIdx], {}, pointerType)));
+
+    // If we are not growing the last memory, then we need to copy data,
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:251]: I would expect to see the real `memory.grow` before copying the data.

File: src/passes/MultiMemoryLowering.cpp [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> adjustOffsetGlobals(Index memIdx) {
+    Builder builder(*wasm);
+    Name functionName =
+      Names::getValidFunctionName(*wasm, "adjust_memory_offsets");
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:239]: It would be good to explicitly put the memory index into the function name. Right now the auto-generated indices appended to the names are off by one.

**Comment by tlively** [comment]
Oops, I hadn't seen @kripken's comments in the meantime. It looks like there's a good amount of overlap. Sorry about that!

File: test/lit/passes/multi-memory-lowering.wast (hunk 1 above)
Comment (ashleynh) [reply test/lit/passes/multi-memory-lowering.wast:12]: Talked offline

File: src/passes/MultiMemoryLowering.cpp [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        assert(iter != parent.memoryIdxMap.end());
+        // No need to change the offset for the first memory
+        if (iter->second == 0) {
+          return;
+        }
+        // There is no offset global for the first memory, so we need to
+        // subtract one when indexing into the offsetGlobalName vector
+        Index idx = iter->second - 1;
+        curr->ptr = builder.makeGlobalGet(parent.offsetGlobalNames[idx],
+                                          parent.pointerType);
```
[the remark this answers is not in the document]
Comment (ashleynh) [reply src/passes/MultiMemoryLowering.cpp:119]: Great catch, thanks

File: src/passes/MultiMemoryLowering.cpp [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+        replaceCurrent(builder.makeCall(funcName, {}, curr->type));
+      }
+
+      // We diverge from the spec here and are not trapping if the offset + type
+      // / 8 is larger than the length of the memory's data Warning,
+      // out-of-bounds loads and stores can read junk out of or corrupt other
+      // memories instead of trapping
+      void visitLoad(Load* curr) {
+        curr->memory = parent.combinedMemory;
+        auto iter = parent.memoryIdxMap.find(curr->memory);
```
[the remark this answers is not in the document]
Comment (ashleynh) [reply src/passes/MultiMemoryLowering.cpp:109]: Great catch, thanks!

File: src/passes/MultiMemoryLowering.cpp [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+          pointerType == Type::i32 ? AddInt32 : AddInt64,
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          dataSegment->offset);
+      }
+    });
+  }
+
+  void createMemorySizeFunctions() {
+    // Don't create a memory size function for the last memory in the vector as
+    // its the combinedMemory we just added
```
[the remark this answers is not in the document]
Comment (ashleynh) [reply src/passes/MultiMemoryLowering.cpp]: Used helper functions to avoid subtracting 1 in various places

File: src/passes/MultiMemoryLowering.cpp [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+  // for each memory
+  std::vector<Name> memorySizeNames;
+  // A vector of the memory grow functions that were created proactively for
+  // each memory
+  std::vector<Name> memoryGrowNames;
+
+  void run(Module* module) override {
+    // Disabling multi-memories so that the target features esction in the
+    // emitted module does not report the use of MultiMemories. Also prevents
+    // later passes from adding additional memories.
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:63]: Maybe move this comment to the top comment at the start of the file? It's important for people using the pass to know it I think.

File: src/passes/MultiMemoryLowering.cpp [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+        auto idx = parent.memoryIdxMap.at(curr->memory);
+        Name funcName = parent.memoryGrowNames[idx];
+        replaceCurrent(builder.makeCall(funcName, {}, curr->type));
+      }
+
+      // We diverge from the spec here and are not trapping if the offset +
+      // write size is larger than the length of the memory's data. Warning:
+      // out-of-bounds loads and stores can read junk out of or corrupt other
+      // memories instead of trapping.
+      // TODO: Add an option to add bounds checks.
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:118]: Also might be best in the top comment at the start.

File: src/passes/MultiMemoryLowering.cpp [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+
+  void addOffsetGlobals() {
+    auto addGlobal = [&](Name name, size_t offset) {
+      auto global =
+        Builder::makeGlobal(name,
+                            pointerType,
+                            Builder(*wasm).makeConst(
+                              pointerType == Type::i32
+                                ? Literal::makeFromInt32(offset, pointerType)
+                                : Literal::makeFromInt64(offset, pointerType)),
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:213]:
```suggestion
                              Literal::makeFromInt64(offset, pointerType)),
```
This works out: we create something of the right type, and if the type is 32, the offset is truncated to 32 bits anyhow, either way. That is, `makeFromInt64` is safe to use to create 32-bit numbers.

File: src/passes/MultiMemoryLowering.cpp [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+                            Builder::Mutable);
+      global->hasExplicitName = true;
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:215]: I think there was another discussion about this for functions - I think we don't need it for globals either.

File: src/passes/MultiMemoryLowering.cpp [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+                         getPageDelta(), combinedMemory, memoryInfo)));
+
+    // If we are not growing the last memory, then we need to copy data,
+    // shifting it over to accomodate the increase from page_delta
+    if (!isLastMemory(memIdx)) {
+      // This offset is the starting pt for copying
+      auto& offsetGlobalName = offsetGlobalNames[memIdx];
+      functionBody = builder.blockify(
+        functionBody,
+        builder.makeMemoryCopy(
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:303]: Really nice how `MemoryCopy` makes this relatively easy... without that this would be much less fun.

**Review by kripken — COMMENTED** [review]
Nice!

Looks correct to me, remaining comments are minor.

It might be useful to manually check this on a wasm-split codebase, for a real-world testcase. That is, using a second memory and then lowering it back down should not cause any runtime errors.

File: src/passes/MultiMemoryLowering.cpp [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+        builder.makeMemorySize(combinedMemory, memoryInfo),
+        builder.makeGlobalGet(offsetGlobalName, pointerType)));
+    } else {
+      auto offsetGlobalName = getOffsetGlobal(memIdx);
+      auto nextOffsetGlobalName = getOffsetGlobal(memIdx + 1);
+      functionBody = builder.makeReturn(builder.makeBinary(
+        Abstract::getBinary(pointerType, Abstract::Sub),
+        builder.makeGlobalGet(nextOffsetGlobalName, pointerType),
+        builder.makeGlobalGet(offsetGlobalName, pointerType)));
+    }
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:372]: There are three cases here, but the test only has two memories so I think we are missing test coverage for one of them.

File: src/passes/MultiMemoryLowering.cpp [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> memoryGrow(Index memIdx) {
+    Builder builder(*wasm);
+    Name name = "custom_memory_grow_" + std::to_string(memIdx);
```
Comment (tlively) [inline src/passes/MultiMemoryLowering.cpp:277]: I don't think the "custom_" prefix adds much here. Maybe we can put the original memory name into the name instead like we do with the offset globals?
```suggestion
    Name name = "memory_grow_" + std::to_string(memIdx);
```

File: test/lit/passes/multi-memory-lowering.wast [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+;; CHECK-NEXT:  (i32.const 1)
+;; CHECK-NEXT: ) "123")
+
+;; CHECK:      (func $custom_memory_size_0 (result i32)
+;; CHECK-NEXT:  (return
+;; CHECK-NEXT:   (global.get $memory2_page_offset)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+
+;; CHECK:      (func $custom_memory_size_1 (result i32)
```
Comment (tlively) [inline test/lit/passes/multi-memory-lowering.wast:32]: It would be good to add a middle memory so we test all three cases for memory_size.

File: test/lit/passes/multi-memory-lowering.wast [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+;; CHECK-NEXT:   )
+;; CHECK-NEXT:  )
+;; CHECK-NEXT:  (memory.copy
+;; CHECK-NEXT:   (i32.add
+;; CHECK-NEXT:    (global.get $memory2_page_offset)
+;; CHECK-NEXT:    (local.get $page_delta)
+;; CHECK-NEXT:   )
+;; CHECK-NEXT:   (global.get $memory2_page_offset)
+;; CHECK-NEXT:   (i32.sub
+;; CHECK-NEXT:    (memory.size)
```
Comment (tlively) [inline test/lit/passes/multi-memory-lowering.wast:58]: This size isn't correct anymore since we have already grown the underlying memory.

File: src/passes/MultiMemoryLowering.cpp [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+             "TODO: handle non-const segment offsets");
+      auto idx = memoryIdxMap.at(dataSegment->memory);
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (idx != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = getOffsetGlobal(idx);
+        dataSegment->offset = builder.makeBinary(
+          Abstract::getBinary(pointerType, Abstract::Add),
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:248]: I believe an add in a data segment offset requires the ExtendedConst feature, so we should error if `!wasm->features.hasExtendedConst()`

File: src/passes/MultiMemoryLowering.cpp (hunk 14 above)
Comment (ashleynh) [reply src/passes/MultiMemoryLowering.cpp:63]: Good idea, thanks!

File: src/passes/MultiMemoryLowering.cpp (hunk 20 above)
Comment (ashleynh) [reply src/passes/MultiMemoryLowering.cpp]: Played around with a few combinations for name, I think memoryName + "_grow" is the most readable. dropping the memidx. Doing the same naming scheme for the replacement memory size function names.

File: test/lit/passes/multi-memory-lowering.wast (hunk 22 above)
Comment (ashleynh) [reply test/lit/passes/multi-memory-lowering.wast:58]: Great catch, thanks!

File: test/lit/passes/multi-memory-lowering.wast [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+;; CHECK-NEXT:  (return
+;; CHECK-NEXT:   (global.get $memory2_page_offset)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+
+;; CHECK:      (func $memory2_size (result i32)
+;; CHECK-NEXT:  (return
+;; CHECK-NEXT:   (i32.sub
+;; CHECK-NEXT:    (global.get $memory3_page_offset)
+;; CHECK-NEXT:    (global.get $memory2_page_offset)
```
Comment (kripken) [inline test/lit/passes/multi-memory-lowering.wast:115]: This looks wrong - `memory.size` returns a value in page size units, so I think this needs to be divided by the page size.

`memory.grow` also takes page sizes. I think the only place that needs the actual byte offsets are loads and stores.

**Review by kripken — COMMENTED** [review]
Great! lgtm % the issue in the comment

File: test/lit/passes/multi-memory-lowering.wast (hunk 24 above)
Comment (ashleynh) [reply test/lit/passes/multi-memory-lowering.wast:115]: Spoke with Alon offline and explained that the (memoryName)_page_offset globals created for each memory are in page units. As a result, the load/store offsets need to change, with the page_offset being multiplied by the page size to get the byte offset that will be added to the existing load/store offset.

Given that we expect there to be far more load/store instructions than memory.grow/ memory.size, I’ll change the (memoryName)_page_offset globals to instead store the byte offset so that we’re able to just conduct a single add operation in each load/store instruction. For memory.grow/memory.size, I’ll divide the byte offset by the page size to get the page units needed for these instructions.

**Review by tlively — COMMENTED** [review]
LGTM % that issue as well 👍

File: src/passes/MultiMemoryLowering.cpp [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+    Expression* functionBody;
+    Index sizeLocal = -1;
+
+    Index returnLocal =
+      Builder::addVar(function.get(), "return_size", pointerType);
+    functionBody = builder.blockify(builder.makeLocalSet(
+      returnLocal, builder.makeCall(memorySizeNames[memIdx], {}, pointerType)));
+
+    if (!isLastMemory(memIdx)) {
+      pageSizeLocal = Builder::addVar(function.get(), "page_size", Type::i32);
```
Comment (kripken) [inline src/passes/MultiMemoryLowering.cpp:301]: A page size local can be avoided, as the `makeConst` can just be done wherever the `local.get` of the local is atm. (There isn't much cost to having multiple appearances of it, too, as it is just a constant. Other things might be different if they take work at runtime though, like an add.)

**Review by kripken — APPROVED** [review]
Great!

lgtm % last comment
